### PR TITLE
Add common versioning script

### PR DESCRIPTION
**Pull request by thardeck**
https://github.com/cloudfoundry-incubator/kubecf/issues/1013

### CODE REVIEW

File: versioning/spec/versioning.rb [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      end
+    end
+
+    context 'when current is not a git dir' do
+      it 'raises an error' do
+        Dir.mktmpdir do |dir|
+          Dir.chdir(dir) do
+            expect {
+              Versioning.verify_git!
+            }.to raise_error(StandardError, /The current directory `#{dir}` is not a git tree!/)
```
Comment (jandubois) [inline versioning/spec/versioning.rb]: The directory returned by `mktmpdir` may include symlinks that `git` will resolve before generating the error message. E.g. on macOS `/tmp` is just a symlink to `/private/tmp`, but `mktmpdir` returns paths relative to `/tmp`.

```suggestion
            }.to raise_error(StandardError, /The current directory `#{Dir.getwd}` is not a git tree!/)
```

**Comment by jandubois** [comment]
I would like to see a specification of what `current_version` is supposed to mean. Is it related to the currently checked out commit? Related to the current branch? The latest commit on the current branch?

What is the actual meaning of that version number, especially the meaning of the `number of commits since tag` thing?
That number would only make sense when the current `HEAD` is actually a child of the latest semver tag. Otherwise it calculates the number of commits since the last shared commit between the current branch and the release branch for the latest tagged release.

E.g. in kubecf the release tags don't live on the `master` branch:

```
$ ruby ../kubecf-tools/versioning/versioning.rb
2.2.2-462.gf4d6761a
$ g rev-list v2.2.2..HEAD --count
462
$ g rev-list v2.2.1..HEAD --count
462
$ g rev-list v2.2.0..HEAD --count
462
```

So this count is only really meaningful if you also include the current branch name in the version tag as well because it only helps with sorting for versions produced from the same branch, but not e.g. for versions numbers created from different PR branches.

Anyways, it is hard to figure out what problem we are trying to solve unless we have a spec that declares what the version number is supposed to represent.

**Comment by jandubois** [comment]
Also, given that this is a brand new repo, can you start with the `master` branch renamed to `main` instead? This will become the default for `git init` in the future, and all CF repos are going to migrate to that naming.

**Comment by viovanov** [comment]
> I would like to see a specification of what `current_version` is supposed to mean. Is it related to the currently checked out commit? Related to the current branch? The latest commit on the current branch?

It's related to the currently checked out commit.
@thardeck if the current commit is tagged (i.e. there are 0 commits since the tag), we shouldn't include the pre-release part.

> 
> What is the actual meaning of that version number, especially the meaning of the `number of commits since tag` thing?
> That number would only make sense when the current `HEAD` is actually a child of the latest semver tag. Otherwise it calculates the number of commits since the last shared commit between the current branch and the release branch for the latest tagged release.
> 

The number of commits and the git sha should be indicators of a pre-release version, as [defined by semver](https://semver.org/#spec-item-9).
The number of commits is human readable (giving the user an idea of how their version relates to the latest common release).
The git hash makes the version unique.

The branch name we're on shouldn't matter.

> E.g. in kubecf the release tags don't live on the `master` branch:
> 
> ```
> $ ruby ../kubecf-tools/versioning/versioning.rb
> 2.2.2-462.gf4d6761a
> $ g rev-list v2.2.2..HEAD --count
> 462
> $ g rev-list v2.2.1..HEAD --count
> 462
> $ g rev-list v2.2.0..HEAD --count
> 462
> ```
> 
> So this count is only really meaningful if you also include the current branch name in the version tag as well because it only helps with sorting for versions produced from the same branch, but not e.g. for versions numbers created from different PR branches.
> 
> Anyways, it is hard to figure out what problem we are trying to solve unless we have a spec that declares what the version number is supposed to represent.

**Comment by thardeck** [comment]
@jandubois many thanks for your feedback, I will create a new pr with all these fixes against the `main` brainch

**Pull request closed without merging**